tfexec/add.go

```diff
@@ -0,0 +1,101 @@
+package tfexec
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os/exec"
+	"strconv"
+)
+
+type addConfig struct {
+	fromState       bool
+	out             string
+	includeOptional bool
+	provider        string
+	reattachInfo    ReattachInfo
+}
+
+var defaultAddOptions = addConfig{}
+
+type AddOption interface {
+	configureAdd(*addConfig)
+}
+
+func (opt *FromStateOption) configureAdd(conf *addConfig) {
+	conf.fromState = opt.fromState
+}
+
+func (opt *OutOption) configureAdd(conf *addConfig) {
+	conf.out = opt.path
+}
+
+func (opt *IncludeOptionalOption) configureAdd(conf *addConfig) {
+	conf.includeOptional = opt.includeOptional
+}
+
+func (opt *ProviderOption) configureAdd(conf *addConfig) {
+	conf.provider = opt.provider
+}
+
+func (opt *ReattachOption) configureAdd(conf *addConfig) {
+	conf.reattachInfo = opt.info
+}
+
+// Add represents the `terraform add` subcommand (added in 1.1.0).
+//
+// Note that this function signature and behaviour is subject
+// to breaking changes including removal of that function
+// until final 1.1.0 Terraform version (with this command) is released.
+func (tf *Terraform) Add(ctx context.Context, address string, opts ...AddOption) (string, error) {
+	cmd, err := tf.addCmd(ctx, address, opts...)
+	if err != nil {
+		return "", err
+	}
+
+	var outBuf bytes.Buffer
```

Comment:
Nitpick: Given that we're returning `string` and never deal with raw bytes we may as well make this [`strings.Builder`](https://pkg.go.dev/strings#Builder)
```suggestion
	var outBuf strings.Builder
```

This admittedly applies to most current `bytes` usage in other places, but that can be addressed in a separate PR.